site/learn/expert-advice/security/breached-password-detection.md

```diff
@@ -0,0 +1,155 @@
+
+## Download and store the breached password datasets
+
+Now that you've catalogued the common passwords and known compromised credentials lists, you can build a system to download and store these data sets. Then you'll want to make them available to your authentication system. 
+
+This is a fairly standard data ingestion problem which has been solved in many contexts, so I won't go into nitty gritty details on how to accomplish this. However, make sure you're ingesting this data on a regular schedule. You'll also want to expose this service to your application or applications, possibly using a REST API or data export.
+
+Some services, such as the aforementioned Have I Been Pwned, also offer an HTTP API. Use these APIs if it fits your application needs. In general storing the datasets will give you more control and flexibility. Like any software system, in building breached password detection, you will need to make tradeoffs. 
+
+It is certainly simpler to rely on a REST API to perform breached password checks. What you lose is:
+
+* *Control*: You can no longer add your own suggested dictionaries or other customizations. It becomes more difficult to isolate your systems from the Internet.
+* *Performance*: You're now going over the Internet, which will almost certainly negatively affect response times. How much? It depends. You'll need to test.
+* *Reliability*: Integrating an outside service which may fail could affect users' ability to authenticate. Plan for the external service to fail to return results in a timely manner and ensure that your authentication system can handle that.
+
+Test any API thoroughly to ensure that your application won't be negatively affected by relying on it. Authentication is a critical part of any application. If it is degraded, then the user experience is degraded as well. 
+
+However, for a quick solution, start with an API integration, perhaps using a [library such as these](https://haveibeenpwned.com/API/Consumers). It's certainly better to use an API integration than to not build this feature. Later, you can build your own data ingestion system as your use increases or you need more control. 
+
+In any event, you'll have a source of compromised credentials and a way to check to see if a password is in that set.
+
+## Configuring your authentication system to check for leaked passwords
+
+Now you'll need to hook into your user management service. Depending on what kind of auth system you have, the integration may be more or less difficult. However, you'll want to check passwords at a couple of different times during the lifecycle of a user's interactions with your systems:
+
+* When they register or an account is created for them.
+* When a user changes a password.
+* When a user logs in.
+* Periodically checking against all users.
```

Comment:
This last one is technically impossible since passwords aren't stored plaintext and checking against breached sets requires plaintext. Plus, the performance implications would be brutal.

Reply:
Ah, I wondered why FusionAuth didn't implement this.

If you're comparing shas, it might be possible, right? Wouldn't you just hash up your plaintext list and do a database query. Look up the hash for user A's password against the corpus?

Agree about performance. That's why I suggested doing it out of band. 

What am I missing/not understanding?

Reply:
salt. I was missing salt.